### ISSUE TITLE
Ingress controller: invalid Envoy configuration created when no ingresses exists

### ISSUE THREAD

**Issue by zcahana** [istio/manager#307]
```shriram@shrirams-mbp:~/istio.io/istio/demos $ kubectl logs istio-ingress-controller-522183506-tff0w
I0313 16:18:45.074407       1 root.go:50] Root flags: &cmd.Flags{Kubeconfig:"", Namespace:""}
I0313 16:18:45.159199       1 client.go:169] Resource already exists: "istio-config.istio.io"
I0313 16:18:45.159224       1 client.go:193] Checking for TPR resources
I0313 16:18:45.161645       1 client.go:203] TPR "IstioConfig" is not ready (the server could not find the requested resource). Waiting...
I0313 16:18:46.171434       1 client.go:203] TPR "IstioConfig" is not ready (the server could not find the requested resource). Waiting...
I0313 16:18:47.174678       1 client.go:203] TPR "IstioConfig" is not ready (the server could not find the requested resource). Waiting...
{
  "listeners": [
    {
      "port": 80,
      "filters": [
        {
          "type": "read",
          "name": "http_connection_manager",
          "config": {
            "codec_type": "auto",
            "stat_prefix": "http",
            "route_config": {},
            "filters": [
              {
                "type": "decoder",
                "name": "router",
                "config": {}
              }
            ],
            "access_log": [
              {
                "path": "/dev/stdout"
              }
            ]
          }
        }
      ],
      "bind_to_port": true
    }
  ],
  "admin": {
    "access_log_path": "/dev/stdout",
    "port": 5000
  },
  "cluster_manager": {
    "sds": {
      "cluster": {
        "name": "sds",
        "connect_timeout_ms": 1000,
        "type": "strict_dns",
        "lb_type": "round_robin",
        "hosts": [
          {
            "url": "tcp://istio-route-controller:8080"
          }
        ]
      },
      "refresh_delay_ms": 1000
    }
  }
}I0313 16:18:48.179242       1 agent.go:139] Envoy starting: [-c /etc/envoy/envoy-rev0.json --restart-epoch 0 --drain-time-s 30 --parent-shutdown-time-s 45 --service-cluster istio-proxy --service-node .]
[2017-03-13 16:18:48.190][14][warning][main] initializing epoch 0 (hot restart version=4.2490504)
[2017-03-13 16:18:48.191][14][critical][main] error initializing configuration '/etc/envoy/envoy-rev0.json': JSON object doesn't conform to schema.
 Invalid schema: #.
 Invalid keyword: required.
 Invalid document key: #
I0313 16:18:48.192913       1 agent.go:187] Envoy epoch 0 terminated: exit status 1
I0313 16:18:48.540974       1 controller.go:144] Event add: key "default/istio-ingress-controller"
I0313 16:18:48.640916       1 controller.go:144] Event add: key "default/kubernetes"
I0313 16:18:48.740951       1 controller.go:144] Event add: key "kube-system/kubernetes-dashboard"
I0313 16:18:48.840997       1 controller.go:144] Event add: key "kube-system/kube-dns"```
```

Zvi Cahana [6:24 PM] 
hmm, empty `route_config`?

gihanson [6:25 PM] 
Zvi might be right - pretty sure routeconfig or rds must be set in the envoy json

[6:25]  
for http_connection_manager

Zvi Cahana [6:29 PM] 
yes, and according to docs, `routeconfig` must have `virtual_hosts` within.

[6:30]  
Solution is probably to remove the `omitempty` in here: https://github.com/istio/manager/blob/master/proxy/envoy/resources.go#L191
GitHub
istio/manager
The Istio manager implementation
 

[6:32]  
I'm assuming Envoy allows for such a configuration, and handles incoming requests to this listener the same as it handles requests that don't match anything within a non-empty `virtual_hosts`.

**Comment by ijsnellf** [comment]
This bug is still present. Fix #313 does not resolve it.

**Comment by zcahana** [comment]
Seems like `cluster_manager.clusters[]` field is required as well. According to the Envoy's config schema, this could be an empty array (correspondingly to the empty `virtual_hosts`), so we just need to remove the `omitempty` from [here](https://github.com/istio/manager/blob/master/proxy/envoy/resources.go#L599).
